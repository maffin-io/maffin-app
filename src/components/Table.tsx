import React from 'react';
import {
  ColumnDef,
  TableOptions,
  useReactTable,
  flexRender,
  getCoreRowModel,
  getSortedRowModel,
  getExpandedRowModel,
  ColumnSort,
  ExpandedState,
  getPaginationRowModel,
} from '@tanstack/react-table';
import classNames from 'classnames';
import { FaSortDown, FaSortUp } from 'react-icons/fa';

import Pagination from '@/components/Pagination';

export type TableProps<T extends object> = {
  id: string,
  columns: ColumnDef<T>[],
  data: T[],
  initialSort?: ColumnSort,
  showHeader?: boolean,
  showPagination?: boolean,
  tdClassName?: string,
  isExpanded?: boolean,
} & Partial<TableOptions<T>>;

export default function Table<T extends object = {}>(
  {
    id,
    initialSort,
    showHeader = true,
    showPagination = false,
    tdClassName = 'px-6 py-4',
    isExpanded = false,
    ...props
  }: TableProps<T>,
): JSX.Element {
  const tableConfig: TableOptions<T> = {
    getCoreRowModel: getCoreRowModel(),
    getSortedRowModel: getSortedRowModel(),
    getExpandedRowModel: getExpandedRowModel(),
    enableExpanding: true,
    initialState: {
      ...props.initialState,
      sorting: (initialSort && [initialSort]) || undefined,
      expanded: isExpanded as ExpandedState,
    },
    ...props,
  };

  if (showPagination && !props.manualPagination) {
    tableConfig.getPaginationRowModel = getPaginationRowModel();
  }

  const table = useReactTable(tableConfig);

  return (
    <>
      <div className="relative overflow-hidden rounded-md">
        <table id={id} className="w-full text-sm text-left">
          {
            showHeader
            && (
              <thead className="bg-white dark:bg-dark-700">
                {table.getHeaderGroups().map((headerGroup) => (
                  <tr key={headerGroup.id}>
                    {headerGroup.headers.map(header => (
                      <th
                        key={header.id}
                        scope="col"
                        className="px-6 py-3"
                        onClick={header.column.getToggleSortingHandler()}
                      >
                        <div className="flex">
                          {flexRender(header.column.columnDef.header, header.getContext())}
                          {(
                            header.column.columnDef.header !== ''
                              && header.column.getCanSort()
                              && (
                                <div className="flex items-center cursor-pointer">
                                  <FaSortUp
                                    className={
                                      classNames(
                                        {
                                          'opacity-100': header.column.getIsSorted() === 'asc',
                                          'opacity-50': header.column.getIsSorted() !== 'asc',
                                        },
                                        'absolute',
                                      )
                                    }
                                  />
                                  <FaSortDown
                                    className={
                                      classNames(
                                        {
                                          'opacity-100': header.column.getIsSorted() === 'desc',
                                          'opacity-50': header.column.getIsSorted() !== 'desc',
                                        },
                                        'absolute',
                                      )
                                    }
                                  />
                                </div>
                              )
                          )}
                        </div>
                      </th>
                    ))}
                  </tr>
                ))}
              </thead>
            )
          }

          <tbody>
            {table.getRowModel().rows.map(row => (
              <tr key={row.id} className="border-b border-white/70 dark:border-dark-700">
                {row.getVisibleCells().map(cell => (
                  <td key={cell.id} className={tdClassName}>
                    {flexRender(cell.column.columnDef.cell, cell.getContext())}
                  </td>
                ))}
              </tr>
            ))}
          </tbody>
        </table>
      </div>

      {
        showPagination
        && <Pagination<T> table={table} />
      }
    </>
  );
}
