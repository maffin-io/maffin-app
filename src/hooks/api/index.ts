import { DateTime } from 'luxon';
import { mutate, SWRResponse } from 'swr';
import useSWRImmutable from 'swr/immutable';

import { Commodity } from '@/book/entities';
import { InvestmentAccount } from '@/book/models';
import * as queries from '@/lib/queries';
import type { PriceDBMap } from '@/book/prices';
import type { AccountsMap } from '@/types/book';
import type { Account, Split, Transaction } from '@/book/entities';

export function useStartDate(): SWRResponse<DateTime> {
  const key = '/api/start-date';
  return useSWRImmutable(
    key,
    fetcher(queries.getEarliestDate, key),
  );
}

export function useMainCurrency(): SWRResponse<Commodity> {
  const key = '/api/main-currency';
  return useSWRImmutable(
    key,
    fetcher(queries.getMainCurrency, key),
  );
}

export function useCommodity(guid: string): SWRResponse<Commodity> {
  const key = `/api/commodities/${guid}`;

  const result = useSWRImmutable(
    key,
    fetcher(() => Commodity.findOneByOrFail({ guid }), key),
  );

  if (result.error) {
    throw new Error(result.error);
  }

  return result;
}

export function useCommodities(): SWRResponse<Commodity[]> {
  const key = '/api/commodities';

  const result = useSWRImmutable(
    key,
    fetcher(() => Commodity.find(), key),
  );

  if (result.error) {
    throw new Error(result.error);
  }

  if (result.data) {
    result.data.forEach(
      (commodity: Commodity) => mutate(
        `/api/commodities/${commodity.guid}`,
        commodity,
        { revalidate: false },
      ),
    );
  }

  return result;
}

export function useLatestTxs(): SWRResponse<Transaction[]> {
  const key = '/api/txs/latest';
  return useSWRImmutable(
    key,
    fetcher(queries.getLatestTxs, key),
  );
}

export function useAccount(guid: string): SWRResponse<Account> {
  const key = `/api/accounts/${guid}`;

  const result = useSWRImmutable(
    key,
    fetcher(() => queries.getAccounts(guid), key),
  );

  if (result.error) {
    throw new Error(result.error);
  }

  return result;
}

export function useAccounts(): SWRResponse<AccountsMap> {
  const key = '/api/accounts';

  const result = useSWRImmutable(
    key,
    fetcher(() => queries.getAccounts(), key),
  );

  if (result.error) {
    throw new Error(result.error);
  }

  if (result.data) {
    Object.values(result.data as AccountsMap).forEach(
      (account: Account) => mutate(`/api/accounts/${account.guid}`, account, { revalidate: false }),
    );
  }

  return result;
}

export function useAccountsMonthlyTotals(): SWRResponse<queries.MonthlyTotals> {
  const { data: accounts } = useAccounts();
  const { data: prices } = useSWRImmutable(
    '/api/prices',
    fetcher(() => queries.getPrices({}), '/api/prices'),
  );

  const key = '/api/monthly-totals';
  const result = useSWRImmutable(
    (accounts && prices) ? key : null,
    fetcher(
      () => queries.getMonthlyTotals(accounts as AccountsMap, prices),
      key,
    ),
  );

  return result;
}

export function useInvestment(guid: string): SWRResponse<InvestmentAccount> {
  const key = `/api/investments/${guid}`;

  const result = useSWRImmutable(
    key,
    fetcher(() => queries.getInvestment(guid), key),
  );

  if (result.error) {
    throw new Error(result.error);
  }

  return result;
}

export function useInvestments(): SWRResponse<InvestmentAccount[]> {
  const key = '/api/investments';

  const result = useSWRImmutable(
    key,
    fetcher(() => queries.getInvestments(), key),
  );

  if (result.error) {
    throw new Error(result.error);
  }

  if (result.data) {
    result.data.forEach(
      (investment: InvestmentAccount) => mutate(
        `/api/investments/${investment.account.guid}`,
        investment,
        { revalidate: false },
      ),
    );
  }

  return result;
}

export function useSplits(guid: string): SWRResponse<Split[]> {
  const key = `/api/splits/${guid}`;
  return useSWRImmutable(
    key,
    fetcher(async () => queries.getSplits(guid), key),
  );
}

/**
 * Returns prices for a given commodity
 */
export function usePrices(c: Commodity | undefined): SWRResponse<PriceDBMap> {
  const key = `/api/prices/${c?.guid}`;
  return useSWRImmutable(
    c?.guid ? key : null,
    fetcher(async () => queries.getPrices({ from: c }), key),
  );
}

function fetcher(f: () => Promise<any>, key: string) {
  return async () => {
    const start = performance.now();
    const r = await f();
    const end = performance.now();
    console.log(`${key}: ${end - start}ms`);
    return r;
  };
}
