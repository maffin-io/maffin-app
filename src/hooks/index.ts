export { default as useDataSource, DataSourceContext } from './useDataSource';
export type { DataSourceContextType } from './useDataSource';
export { default as useBookStorage } from './useBookStorage';
export { default as useGapiClient } from './useGapiClient';
