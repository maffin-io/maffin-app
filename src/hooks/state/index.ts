export { useTheme } from './useTheme';
export { useOnline } from './useOnline';
