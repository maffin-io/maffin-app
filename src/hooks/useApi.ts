import useSWRImmutable from 'swr/immutable';
import { Fetcher, SWRResponse, SWRConfiguration } from 'swr';

import { Commodity } from '@/book/entities';
import * as queries from '@/lib/queries';
import { PriceDB } from '@/book/prices';
import getUser from '@/lib/getUser';

/**
 * Wrapper around SWR to make the usable keys explicit and re-usable
 */
export default function useApi(key: ApiPaths | null): SWRResponse {
  const f = key !== null ? API[key].f : () => {};
  const swrArgs = key !== null ? API[key].swrArgs : undefined;

  return useSWRImmutable(key, f, swrArgs);
}

export type ApiPaths = (
  '/api/start-date'
  | '/api/main-currency'
  | '/api/commodities'
  | '/api/accounts'
  | '/api/investments'
  | '/api/prices/today'
  | '/api/txs/latest'
  | '/api/user'
);

const API: {
  [key in ApiPaths]: {
    f: Fetcher,
    swrArgs?: SWRConfiguration,
  };
} = {
  '/api/start-date': {
    f: queries.getEarliestDate,
  },
  '/api/main-currency': {
    f: queries.getMainCurrency,
  },
  '/api/commodities': {
    f: Commodity.find,
  },
  '/api/accounts': {
    f: async (k: string) => {
      const start = performance.now();
      const r = await queries.getAccounts();
      const end = performance.now();
      console.log(`${k}: ${end - start}ms`);
      return r;
    },
  },
  '/api/investments': {
    f: queries.getInvestments,
  },
  '/api/prices/today': {
    f: PriceDB.getTodayQuotes,
  },
  '/api/txs/latest': {
    f: queries.getLatestTxs,
  },
  '/api/user': {
    f: getUser,
    swrArgs: {
      refreshInterval: 100000,
      revalidateOnMount: true,
    },
  },
};
