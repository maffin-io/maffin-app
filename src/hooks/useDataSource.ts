import React from 'react';
import { DataSource } from 'typeorm';
import initSqlJs from 'sql.js';
import { mutate } from 'swr';
import pako from 'pako';

import { insertTodayPrices } from '@/lib/Stocker';
import useBookStorage from '@/hooks/useBookStorage';
import { importBook as importGnucashBook } from '@/lib/gnucash';
import {
  Account,
  Book,
  Commodity,
  Price,
  Split,
  Transaction,
} from '@/book/entities';
import { MIGRATIONS } from '@/book/migrations';
import { isStaging } from '@/helpers/env';
import { useQueryClient } from '@tanstack/react-query';
import type BookStorage from '@/lib/storage/BookStorage';

export type DataSourceContextType = {
  datasource: DataSource | null,
  save: () => Promise<void>,
  importBook: (rawBook: Uint8Array) => Promise<void>,
  isLoaded: boolean,
};

const DATASOURCE: DataSource = new DataSource({
  type: 'sqljs',
  synchronize: true,
  logging: false,
  entities: [Account, Book, Commodity, Price, Split, Transaction],
  migrations: MIGRATIONS,
});

export const DataSourceContext = React.createContext<DataSourceContextType>({
  datasource: DATASOURCE,
  save: async () => {},
  importBook: async () => {},
  isLoaded: false,
});

export default function useDataSource(): DataSourceContextType {
  const { storage } = useBookStorage();
  const [isLoaded, setIsLoaded] = React.useState(DATASOURCE.isInitialized);
  const queryClient = useQueryClient();

  React.useEffect(() => {
    async function load() {
      let rawBook: Uint8Array;
      if (storage && !DATASOURCE.isInitialized && queryClient) {
        ([, rawBook] = await Promise.all([
          initSqlJs({ locateFile: file => `/${file}` }),
          storage.get(),
        ]));

        // Due to async nature, this value can change between this and the previous statement
        // as it's a shared global
        if (!DATASOURCE.isInitialized) {
          DATASOURCE.setOptions({
            type: 'sqljs',
            synchronize: true,
            logging: false,
            entities: [Account, Book, Commodity, Price, Split, Transaction],
            migrations: MIGRATIONS,
            extra: {
              queryClient,
            },
          });
          const start = performance.now();
          await DATASOURCE.initialize();

          if (rawBook.length) {
            await DATASOURCE.sqljsManager.loadDatabase(rawBook);
            await DATASOURCE.runMigrations();
          } else {
            await createEmptyBook();
          }
          const end = performance.now();
          console.log(`init datasource: ${end - start}ms`);
        }
        await loadStockerPrices();
        setIsLoaded(true);
      }
    }

    load();
  }, [storage, queryClient]);

  if (!isLoaded) {
    return {
      datasource: DATASOURCE,
      save: async () => {},
      importBook: async () => {},
      isLoaded: false,
    };
  }

  return {
    datasource: DATASOURCE,
    save: () => save(storage as BookStorage),
    importBook: (rawData: Uint8Array) => importBook(storage as BookStorage, rawData),
    isLoaded,
  };
}

async function save(storage: BookStorage) {
  mutate('/state/save', true, { revalidate: false });
  const rawBook = DATASOURCE.sqljsManager.exportDatabase();
  await storage.save(rawBook);
  mutate('/state/save', false, { revalidate: false });
}

async function importBook(storage: BookStorage, rawData: Uint8Array) {
  let parsedData;
  try {
    parsedData = pako.ungzip(rawData);
  } catch (err) {
    parsedData = rawData;
  }

  const rawBook = await importGnucashBook(parsedData);
  await DATASOURCE.sqljsManager.loadDatabase(rawBook);

  DATASOURCE.options.extra.queryClient.refetchQueries({
    type: 'all',
  });

  await loadStockerPrices();
  await save(storage);
}

async function createEmptyBook() {
  const rootAccount = Account.create({
    guid: 'rootAccount',
    name: 'Root',
    type: 'ROOT',
  });
  await Account.upsert(rootAccount, ['guid']);

  await Book.upsert(
    {
      guid: 'maffinBook',
      fk_root: rootAccount.guid,
    },
    ['guid'],
  );
}

async function loadStockerPrices() {
  if (!isStaging()) {
    try {
      // We have to await because we use Price.upsert. Once we correct this,
      // saving Price already updates the local cache so we don't need await
      await insertTodayPrices();
    } catch (e) {
      console.warn(`Retrieving live prices failed ${e}`);
    }
  }
}
