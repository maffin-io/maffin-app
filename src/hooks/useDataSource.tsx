import React from 'react';
import { DataSource, QueryFailedError } from 'typeorm';
import initSqlJs from 'sql.js';
import { Settings } from 'luxon';
import pako from 'pako';

import { insertTodayPrices } from '@/lib/prices';
import useBookStorage from '@/hooks/useBookStorage';
import { migrate as migrateFromGnucash } from '@/lib/gnucash';
import {
  Account,
  Book,
  Commodity,
  Price,
  Split,
  Transaction,
} from '@/book/entities';
import { MIGRATIONS } from '@/book/migrations';
import { useQueryClient } from '@tanstack/react-query';
import type BookStorage from '@/lib/storage/BookStorage';
import { MaffinError, StorageError } from '@/helpers/errors';

export type DataSourceContextType = {
  datasource: DataSource | null,
  save: () => Promise<void>,
  importBook: (rawBook: Uint8Array) => Promise<void>,
  isLoaded: boolean,
};

const DATASOURCE: DataSource = new DataSource({
  type: 'sqljs',
  synchronize: true,
  logging: false,
  entities: [Account, Book, Commodity, Price, Split, Transaction],
  migrations: MIGRATIONS,
});

export const DataSourceContext = React.createContext<DataSourceContextType>({
  datasource: DATASOURCE,
  save: async () => {},
  importBook: async () => {},
  isLoaded: false,
});

export default function useDataSource(): DataSourceContextType {
  const { storage } = useBookStorage();
  const [isLoaded, setIsLoaded] = React.useState(DATASOURCE.isInitialized);
  const queryClient = useQueryClient();

  React.useEffect(() => {
    async function load() {
      let rawBook: Uint8Array;
      if (storage && !DATASOURCE.isInitialized && queryClient) {
        ([, rawBook] = await Promise.all([
          initSqlJs({ locateFile: file => `/${file}` }),
          storage.get(),
        ]));

        // Due to async nature, this value can change between this and the previous statement
        // as it's a shared global
        if (!DATASOURCE.isInitialized) {
          DATASOURCE.setOptions({
            type: 'sqljs',
            synchronize: true,
            logging: false,
            entities: [Account, Book, Commodity, Price, Split, Transaction],
            migrations: MIGRATIONS,
            extra: {
              queryClient,
            },
          });
          const start = performance.now();
          await DATASOURCE.initialize();

          if (rawBook.length) {
            await DATASOURCE.sqljsManager.loadDatabase(rawBook);
            await DATASOURCE.runMigrations();
          } else {
            await createEmptyBook();
          }
          const end = performance.now();
          console.log(`init datasource: ${end - start}ms`);
        }
        loadStockerPrices();
        setIsLoaded(true);
      }
    }

    load();
  }, [storage, queryClient]);

  if (!isLoaded) {
    return {
      datasource: DATASOURCE,
      save: async () => {},
      importBook: async () => {},
      isLoaded: false,
    };
  }

  return {
    datasource: DATASOURCE,
    save: () => save(storage as BookStorage),
    importBook: (rawData: Uint8Array) => importBook(storage as BookStorage, rawData),
    isLoaded,
  };
}

async function save(storage: BookStorage) {
  DATASOURCE.options.extra.queryClient.setQueryData(['state', 'isSaving'], true);
  await DATASOURCE.query('VACUUM');
  const rawBook = DATASOURCE.sqljsManager.exportDatabase();
  try {
    await storage.save(rawBook);
  } catch (e) {
    (e as MaffinError).show();
  } finally {
    DATASOURCE.options.extra.queryClient.setQueryData(['state', 'isSaving'], false);
  }
}

async function importBook(storage: BookStorage, rawData: Uint8Array) {
  if (process.env.NEXT_PUBLIC_ENV !== 'master') {
    Settings.now = () => Date.now();
  }
  let parsedData;
  try {
    parsedData = pako.ungzip(rawData);
  } catch (err) {
    parsedData = rawData;
  }

  const previousData = DATASOURCE.sqljsManager.exportDatabase();
  await DATASOURCE.sqljsManager.loadDatabase(parsedData);
  try {
    // Sadly loadDatabase doesn't check that the file is valid so we call this to trigger
    // errors if any
    await DATASOURCE.query('VACUUM');
  } catch (e) {
    let code = 'UNKNOWN';
    if (e instanceof QueryFailedError) {
      code = 'INVALID_FILE';
    }
    new StorageError((e as Error).message, code).show();
    await DATASOURCE.sqljsManager.loadDatabase(previousData);
    return;
  }

  // We probably should do this conditionally only for when it is a gnucash file
  await migrateFromGnucash();
  DATASOURCE.options.extra.queryClient.refetchQueries({
    queryKey: ['api'],
    type: 'all',
  });

  loadStockerPrices();
  await save(storage);
}

async function createEmptyBook() {
  const rootAccount = Account.create({
    guid: 'rootAccount',
    name: 'Root',
    type: 'ROOT',
  });
  await Account.upsert(rootAccount, ['guid']);

  await Book.upsert(
    {
      guid: 'maffinBook',
      fk_root: rootAccount.guid,
    },
    ['guid'],
  );
}

async function loadStockerPrices() {
  try {
    // We have to await because we use Price.upsert. Once we correct this,
    // saving Price already updates the local cache so we don't need await
    await insertTodayPrices();
  } catch (e) {
    console.warn(`Retrieving live prices failed ${e}`);
  }
}
