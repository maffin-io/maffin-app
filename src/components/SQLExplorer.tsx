import React from 'react';
import Table from '@/components/tables/Table';
import { BiSolidDownArrow, BiSolidRightArrow } from 'react-icons/bi';
import type { ColumnMetadata } from 'typeorm/metadata/ColumnMetadata';
import type { PropsWithChildren } from 'react';

import { useDataSource } from '@/hooks';
import Loading from './Loading';

export default function SQLExplorer(): React.JSX.Element {
  const { datasource } = useDataSource();

  if (!datasource) {
    return <Loading />;
  }

  const metadatas = datasource.entityMetadatas;

  return (
    <ul>
      {
        metadatas.map(metadata => (
          <li key={metadata.givenTableName}>
            <Expandable
              content={(
                <span className="text-md">
                  {metadata.givenTableName}
                </span>
              )}
            >
              <Table<ColumnMetadata>
                id={`sql-table-${metadata.givenTableName}`}
                showHeader={false}
                columns={[
                  {
                    accessorKey: 'databaseName',
                  },
                  {
                    id: 'type',
                    accessorFn: (row: ColumnMetadata) => {
                      // @ts-ignore
                      const { name } = row.type;
                      if (name) {
                        return name;
                      }

                      const { length } = row;

                      if (length) {
                        return `${row.type}(${length})`;
                      }

                      return row.type;
                    },
                  },
                ]}
                data={metadata.columns.filter(m => m.isSelect)}
              />
            </Expandable>
          </li>
        ))
      }
    </ul>
  );
}

function Expandable({
  content,
  children,
}: { content: React.JSX.Element } & PropsWithChildren): React.JSX.Element {
  const [expanded, setExpanded] = React.useState(false);

  return (
    <>
      <div
        onClick={() => setExpanded(!expanded)}
        className="flex items-center cursor-pointer"
      >
        {
          expanded
            ? <BiSolidDownArrow className="mr-1" />
            : <BiSolidRightArrow className="mr-1" />
        }
        {content}
      </div>
      <div
        className={`${expanded ? 'visible' : 'hidden'}`}
      >
        {children}
      </div>
    </>
  );
}
