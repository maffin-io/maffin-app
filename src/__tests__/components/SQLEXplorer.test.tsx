import React from 'react';
import { render, screen } from '@testing-library/react';
import { userEvent } from '@testing-library/user-event';
import type { EntityMetadata } from 'typeorm';
import type { ColumnMetadata } from 'typeorm/metadata/ColumnMetadata';

import SQLExplorer from '@/components/SQLExplorer';
import Table from '@/components/tables/Table';
import * as dataSourceHooks from '@/hooks/useDataSource';
import type { DataSourceContextType } from '@/hooks';

jest.mock('@/hooks/useDataSource', () => ({
  __esModule: true,
  ...jest.requireActual('@/hooks/useDataSource'),
}));

jest.mock('@/components/tables/Table', () => jest.fn(
  () => <div data-testid="Table" />,
));

jest.mock('@/components/Loading', () => jest.fn(
  () => <div data-testid="Loading" />,
));

describe('SQLExplorer', () => {
  let metadatas: EntityMetadata[];

  beforeEach(() => {
    metadatas = [
      {
        givenTableName: 'table1',
        columns: [
          {
            databaseName: 'column1',
            type: 'varchar',
            isSelect: true,
          },
        ],
      } as EntityMetadata,
      {
        givenTableName: 'table2',
        columns: [
          {
            databaseName: 'column2',
            type: 'varchar',
            isSelect: true,
          },
        ],
      } as EntityMetadata,
    ];
    jest.spyOn(dataSourceHooks, 'default').mockReturnValue({
      datasource: {
        entityMetadatas: metadatas,
      },
    } as DataSourceContextType);
  });

  afterEach(() => {
    jest.clearAllMocks();
  });

  it('returns loading when no datasource', async () => {
    jest.spyOn(dataSourceHooks, 'default').mockReturnValue({
      datasource: null,
    } as DataSourceContextType);

    render(<SQLExplorer />);
    await screen.findByTestId('Loading');
  });

  it('renders all metadata', () => {
    const { container } = render(<SQLExplorer />);
    expect(container).toMatchSnapshot();
    expect(Table as jest.Mock).nthCalledWith(
      1,
      {
        id: 'sql-table-table1',
        showHeader: false,
        columns: [
          {
            accessorKey: 'databaseName',
          },
          {
            accessorFn: expect.any(Function),
            id: 'type',
          },
        ],
        data: metadatas[0].columns,
      },
      {},
    );
    expect(Table as jest.Mock).nthCalledWith(
      2,
      {
        id: 'sql-table-table2',
        showHeader: false,
        columns: [
          {
            accessorKey: 'databaseName',
          },
          {
            accessorFn: expect.any(Function),
            id: 'type',
          },
        ],
        data: metadatas[1].columns,
      },
      {},
    );
    // eslint-disable-next-line testing-library/no-node-access
    expect(screen.getAllByTestId('Table')[0].parentElement).toHaveClass('hidden');
  });

  it('filters non select columns', () => {
    const metadata = [
      {
        givenTableName: 'table1',
        columns: [
          {
            databaseName: 'column1',
            type: 'varchar',
            isSelect: true,
          },
          {
            databaseName: 'column2',
            type: 'varchar',
            isSelect: false,
          },
        ],
      } as EntityMetadata,
    ];
    jest.spyOn(dataSourceHooks, 'default').mockReturnValue({
      datasource: {
        entityMetadatas: metadata,
      },
    } as DataSourceContextType);

    render(<SQLExplorer />);

    expect(Table as jest.Mock).toBeCalledWith(
      expect.objectContaining({
        data: [metadata[0].columns[0]],
      }),
      {},
    );
  });

  it('picks type', () => {
    const metadata = [
      {
        givenTableName: 'table1',
        columns: [] as ColumnMetadata[],
      } as EntityMetadata,
    ];
    jest.spyOn(dataSourceHooks, 'default').mockReturnValue({
      datasource: {
        entityMetadatas: metadata,
      },
    } as DataSourceContextType);

    render(<SQLExplorer />);

    const { accessorFn } = (Table as jest.Mock).mock.calls[0][0].columns[1];
    expect(accessorFn({ type: 'name' })).toEqual('name');
  });

  it('picks type with length', () => {
    const metadata = [
      {
        givenTableName: 'table1',
        columns: [] as ColumnMetadata[],
      } as EntityMetadata,
    ];
    jest.spyOn(dataSourceHooks, 'default').mockReturnValue({
      datasource: {
        entityMetadatas: metadata,
      },
    } as DataSourceContextType);

    render(<SQLExplorer />);

    const { accessorFn } = (Table as jest.Mock).mock.calls[0][0].columns[1];
    expect(accessorFn({ type: 'name', length: '10' })).toEqual('name(10)');
  });

  it('picks type name', () => {
    const metadata = [
      {
        givenTableName: 'table1',
        columns: [] as ColumnMetadata[],
      } as EntityMetadata,
    ];
    jest.spyOn(dataSourceHooks, 'default').mockReturnValue({
      datasource: {
        entityMetadatas: metadata,
      },
    } as DataSourceContextType);

    render(<SQLExplorer />);

    const { accessorFn } = (Table as jest.Mock).mock.calls[0][0].columns[1];
    expect(accessorFn({ type: { name: 'name' } })).toEqual('name');
  });

  it('displays table on click', async () => {
    const metadata = [
      {
        givenTableName: 'table1',
        columns: [] as ColumnMetadata[],
      } as EntityMetadata,
    ];
    jest.spyOn(dataSourceHooks, 'default').mockReturnValue({
      datasource: {
        entityMetadatas: metadata,
      },
    } as DataSourceContextType);

    render(<SQLExplorer />);

    const title = screen.getByText('table1');
    await userEvent.click(title);

    // eslint-disable-next-line testing-library/no-node-access
    expect(screen.getByTestId('Table').parentElement).toHaveClass('visible');
  });
});
