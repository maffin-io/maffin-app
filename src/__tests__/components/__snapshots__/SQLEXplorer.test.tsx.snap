// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SQLExplorer renders all metadata 1`] = `
<div>
  <ul>
    <li>
      <div
        class="flex items-center cursor-pointer"
      >
        <svg
          class="mr-1"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M5.536 21.886a1.004 1.004 0 0 0 1.033-.064l13-9a1 1 0 0 0 0-1.644l-13-9A1 1 0 0 0 5 3v18a1 1 0 0 0 .536.886z"
          />
        </svg>
        <span
          class="text-md"
        >
          table1
        </span>
      </div>
      <div
        class="hidden"
      >
        <div
          data-testid="Table"
        />
      </div>
    </li>
    <li>
      <div
        class="flex items-center cursor-pointer"
      >
        <svg
          class="mr-1"
          fill="currentColor"
          height="1em"
          stroke="currentColor"
          stroke-width="0"
          viewBox="0 0 24 24"
          width="1em"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M5.536 21.886a1.004 1.004 0 0 0 1.033-.064l13-9a1 1 0 0 0 0-1.644l-13-9A1 1 0 0 0 5 3v18a1 1 0 0 0 .536.886z"
          />
        </svg>
        <span
          class="text-md"
        >
          table2
        </span>
      </div>
      <div
        class="hidden"
      >
        <div
          data-testid="Table"
        />
      </div>
    </li>
  </ul>
</div>
`;
